File: apps/sting-go/android/app/src/main/java/run/stingjs/go/StingGoManifest.kt
package run.stingjs.go

import org.json.JSONObject

data class StingGoManifest(
    val schemaVersion: Int,
    val runtimeVersion: String,
    val engine: String,
    val projectName: String,
    val bundlePath: String,
    val reloadPath: String,
    val healthPath: String,
    val reportPath: String?,
    val capabilities: Set<String>,
) {
    companion object {
        const val SUPPORTED_SCHEMA_VERSION = 1
        const val SUPPORTED_RUNTIME_VERSION = "0.1.0"
        const val SUPPORTED_ENGINE = "quickjs"

        fun parse(source: String): StingGoManifest {
            val json = JSONObject(source)
            val schemaVersion = json.getInt("schemaVersion")
            require(schemaVersion == SUPPORTED_SCHEMA_VERSION) {
                "Unsupported Sting Go manifest schema $schemaVersion; expected $SUPPORTED_SCHEMA_VERSION"
            }

            val runtimeVersion = json.getString("runtimeVersion")
            require(runtimeVersion == SUPPORTED_RUNTIME_VERSION) {
                "Project requires Sting runtime $runtimeVersion; this Sting Go supports $SUPPORTED_RUNTIME_VERSION"
            }

            val engine = json.getString("engine")
            require(engine == SUPPORTED_ENGINE) {
                "Project requires JavaScript engine $engine; this Sting Go uses $SUPPORTED_ENGINE"
            }

            val project = json.getJSONObject("project")
            val bundle = json.getJSONObject("bundle")
            val bundlePath = bundle.getString("path")
            require(bundlePath == "/bundle") { "Unsupported Sting Go bundle path $bundlePath" }
            require(bundle.getString("contentType") == "application/javascript") {
                "Unsupported Sting Go bundle content type"
            }

            val development = json.getJSONObject("development")
            val reload = development.getJSONObject("reload")
            val reloadPath = reload.getString("path")
            require(reloadPath == "/events") { "Unsupported Sting Go reload path $reloadPath" }
            require(reload.getString("transport") == "sse") {
                "Unsupported Sting Go reload transport; v1 requires sse"
            }
            require(reload.getString("contentType") == "text/event-stream") {
                "Unsupported Sting Go reload content type"
            }

            val health = development.getJSONObject("health")
            val healthPath = health.getString("path")
            require(healthPath == "/health") { "Unsupported Sting Go health path $healthPath" }
            require(health.getString("contentType") == "application/json") {
                "Unsupported Sting Go health content type"
            }

            val reportPath = development.optJSONObject("report")?.let { report ->
                val path = report.getString("path")
                require(path == "/report") { "Unsupported Sting Go report path $path" }
                require(report.getString("method") == "POST") {
                    "Unsupported Sting Go report method; v1 requires POST"
                }
                require(report.getString("contentType") == "application/json") {
                    "Unsupported Sting Go report content type"
                }
                path
            }

            val capabilitiesJson = json.getJSONArray("capabilities")
            val capabilities = buildSet {
                for (index in 0 until capabilitiesJson.length()) {
                    add(capabilitiesJson.getString(index))
                }
            }

            return StingGoManifest(
                schemaVersion = schemaVersion,
                runtimeVersion = runtimeVersion,
                engine = engine,
                projectName = project.getString("name"),
                bundlePath = bundlePath,
                reloadPath = reloadPath,
                healthPath = healthPath,
                reportPath = reportPath,
                capabilities = capabilities,
            )
        }
    }
}

File: apps/sting-go/ios/Support/Sources/StingGoSupport/StingGoManifest.swift
import Foundation

public enum StingGoProtocolError: LocalizedError, Equatable {
    case invalidManifest(String)
    case unsupportedSchema(Int)
    case unsupportedRuntime(required: String, supported: String)
    case unsupportedEngine(String)
    case unsupportedBundle(path: String, contentType: String)
    case unsupportedReload(path: String, transport: String, contentType: String)
    case unsupportedHealth(path: String, contentType: String)
    case unsupportedReport(path: String, method: String, contentType: String)
    case unsupportedCapabilities([String])
    case invalidEndpoint(String)

    public var errorDescription: String? {
        switch self {
        case .invalidManifest(let message):
            return "Invalid Sting Go manifest: \(message)"
        case .unsupportedSchema(let version):
            return "Unsupported Sting Go manifest schema \(version); expected 1"
        case .unsupportedRuntime(let required, let supported):
            return "Project requires Sting runtime \(required); this Sting Go supports \(supported)"
        case .unsupportedEngine(let engine):
            return "Project requires JavaScript engine \(engine); this Sting Go uses quickjs"
        case .unsupportedBundle(let path, let contentType):
            return "Unsupported Sting Go bundle endpoint \(path) (\(contentType))"
        case .unsupportedReload(let path, let transport, let contentType):
            return "Unsupported Sting Go reload endpoint \(path) (\(transport), \(contentType))"
        case .unsupportedHealth(let path, let contentType):
            return "Unsupported Sting Go health endpoint \(path) (\(contentType))"
        case .unsupportedReport(let path, let method, let contentType):
            return "Unsupported Sting Go report endpoint \(path) (\(method), \(contentType))"
        case .unsupportedCapabilities(let capabilities):
            return "This Sting Go build does not include required capabilities: \(capabilities.joined(separator: ", "))"
        case .invalidEndpoint(let path):
            return "Could not resolve Sting Go endpoint \(path) relative to the manifest URL"
        }
    }
}

public struct StingGoManifest: Decodable, Equatable, Sendable {
    public struct Project: Decodable, Equatable, Sendable {
        public let name: String
    }

    public struct Endpoint: Decodable, Equatable, Sendable {
        public let path: String
        public let contentType: String
    }

    public struct ReloadEndpoint: Decodable, Equatable, Sendable {
        public let path: String
        public let transport: String
        public let contentType: String
    }

    public struct ReportEndpoint: Decodable, Equatable, Sendable {
        public let path: String
        public let method: String
        public let contentType: String
    }

    public struct Development: Decodable, Equatable, Sendable {
        public let reload: ReloadEndpoint
        public let health: Endpoint
        public let report: ReportEndpoint?
    }

    public static let supportedSchemaVersion = 1
    public static let supportedRuntimeVersion = "0.1.0"
    public static let supportedEngine = "quickjs"

    public let schemaVersion: Int
    public let runtimeVersion: String
    public let engine: String
    public let project: Project
    public let bundle: Endpoint
    public let development: Development
    public let capabilities: Set<String>

    public static func decode(_ data: Data) throws -> StingGoManifest {
        do {
            return try JSONDecoder().decode(StingGoManifest.self, from: data)
        } catch {
            throw StingGoProtocolError.invalidManifest(error.localizedDescription)
        }
    }

    @discardableResult
    public func validate(
        clientRuntimeVersion: String = StingGoManifest.supportedRuntimeVersion,
        clientCapabilities: Set<String>
    ) throws -> StingGoManifest {
        guard schemaVersion == Self.supportedSchemaVersion else {
            throw StingGoProtocolError.unsupportedSchema(schemaVersion)
        }
        guard runtimeVersion == clientRuntimeVersion else {
            throw StingGoProtocolError.unsupportedRuntime(
                required: runtimeVersion,
                supported: clientRuntimeVersion
            )
        }
        guard engine == Self.supportedEngine else {
            throw StingGoProtocolError.unsupportedEngine(engine)
        }
        guard bundle.path == "/bundle", bundle.contentType == "application/javascript" else {
            throw StingGoProtocolError.unsupportedBundle(
                path: bundle.path,
                contentType: bundle.contentType
            )
        }
        let reload = development.reload
        guard reload.path == "/events",
              reload.transport == "sse",
              reload.contentType == "text/event-stream" else {
            throw StingGoProtocolError.unsupportedReload(
                path: reload.path,
                transport: reload.transport,
                contentType: reload.contentType
            )
        }
        let health = development.health
        guard health.path == "/health", health.contentType == "application/json" else {
            throw StingGoProtocolError.unsupportedHealth(
                path: health.path,
                contentType: health.contentType
            )
        }
        if let report = development.report {
            guard report.path == "/report",
                  report.method == "POST",
                  report.contentType == "application/json" else {
                throw StingGoProtocolError.unsupportedReport(
                    path: report.path,
                    method: report.method,
                    contentType: report.contentType
                )
            }
        }

        let unsupported = capabilities.subtracting(clientCapabilities).sorted()
        guard unsupported.isEmpty else {
            throw StingGoProtocolError.unsupportedCapabilities(unsupported)
        }
        return self
    }

    public func endpointURL(path: String, relativeTo manifestURL: URL) throws -> URL {
        guard let url = URL(string: path, relativeTo: manifestURL)?.absoluteURL,
              url.scheme == "http" || url.scheme == "https" else {
            throw StingGoProtocolError.invalidEndpoint(path)
        }
        return url
    }
}

public struct StingGoHealth: Decodable, Equatable, Sendable {
    public let ok: Bool
    public let watching: Bool
    public let reloadVersion: Int
}
